src/hmacKey.ts

```diff
@@ -0,0 +1,355 @@
+      callback!(null, this.metadata, res);
+    });
+  }
+
+  update(
+    metadata: UpdateHmacKeyMetadata,
+    options?: UpdateHmacKeyOptions
+  ): Promise<HmacKeyMetadataResponse>;
+  update(
+    metadata: UpdateHmacKeyMetadata,
+    callback: HmacKeyMetadataCallback
+  ): void;
+  update(
+    metadata: UpdateHmacKeyMetadata,
+    options: UpdateHmacKeyOptions,
+    callback: HmacKeyMetadataCallback
+  ): void;
+  /**
+   * @typedef {object} UpdateHmacKeyMetadata Subset of {@link HmacKeyMetadata} to update.
+   * @property {string} state New state of the HmacKey. Either 'ACTIVE' or 'INACTIVE'.
+   * @property {string} [etag] Include an etag from a previous get HMAC key request
+   *    to perform safe read-modify-write.
+   */
+  /**
+   * @typedef {object} UpdateHmacKeyOptions
+   * @property {string} userProject This parameter is currently ignored.
+   */
+  /**
+   * @callback HmacKeyMetadataCallback
+   * @param {?Error} err Request error, if any.
+   * @param {HmacKeyMetadata} metadata The updated {@link HmacKeyMetadata} object.
+   * @param {object} apiResponse The full API response.
+   */
+  /**
+   * @typedef {array} HmacKeyMetadataResponse
+   * @property {HmacKeyMetadata} 0 The updated {@link HmacKeyMetadata} object.
+   * @property {object} 1 The full API response.
+   */
+  /**
+   * Updates the state of an HMAC key. See {@link UpdateHmacKeyMetadata} for
+   * valid states.
+   *
+   * @param {UpdateHmacKeyMetadata} metadata The new metadata.
+   * @param {UpdateHmacKeyOptions} [options] Configuration options.
+   * @param {HmacKeyMetadataCallback} [callback] Callback function.
+   * @returns {Promise<HmacKeyMetadataResponse>}
+   *
+   * @example
+   * const {Storage} = require('@google-cloud/storage');
+   * const storage = new Storage();
+   *
+   * const metadata = {
+   *   state: 'INACTIVE',
+   * };
+   *
+   * storage.hmacKey('ACCESS_ID')
+   *   .update(metadata, (err, hmacKeyMetadata) => {
+   *     if (err) {
+   *       // The request was an error.
+   *       console.error(err);
+   *       return;
+   *     }
+   *     console.log(hmacKeyMetadata);
+   *   });
+   *
+   * //-
+   * // If the callback is omitted, a promise is returned.
+   * //-
+   * storage.hmacKey('ACCESS_ID')
+   *   .update(metadata)
+   *   .then((data) => {
+   *     const hmacKeyMetadata = data[0];
+   *     console.log(hmacKeyMetadata);
+   *   });
+   */
+  update(
+    metadata: UpdateHmacKeyMetadata,
+    optionsOrCb?: UpdateHmacKeyOptions | HmacKeyMetadataCallback,
+    cb?: HmacKeyMetadataCallback
+  ): Promise<HmacKeyMetadataResponse> | void {
+    if (
+      typeof metadata !== 'object' ||
+      Object.getOwnPropertyNames(metadata).length === 0
+    ) {
+      throw new Error(
+        'Cannot update HmacKey with an undefined/empty options object.'
+      );
+    }
+
+    const {options, callback} = normalize<
+      UpdateHmacKeyOptions,
+      HmacKeyMetadataCallback
+    >(optionsOrCb, cb);
+
+    this.request(
```

Comment:
`this.setMetadata()` should be able to do all of this, and we won't need the `update()` method.

Reply:
ack.

Reply:
hmacKey only provides `PUT` method to update the key, while `setMetadata` makes a `PATCH` request which it doesn't support.

Reply:
You can actually configure this in the constructor:

setMetadata: {
  reqOpts: {
    method: 'PUT',
  },
}

Reply:
To elaborate on that, that object would go in the `const methods = {}` object in the constructor, as a sibling to `get: true` and the other properties. When you need further customization, you can use an object instead of `true` to define default request options.

Reply:
It looks like setting `reqOpts` here will not override the `method` config:

https://github.com/googleapis/nodejs-common/blob/deea6b2c5581b0d9892b98c7b17506039dd1a5e4/src/service-object.ts#L455-L460

The override is overwritten by the last object.

Reply:
[*forehead slap*](https://github.com/googleapis/nodejs-common/pull/353/files)

I can fix this soon, but if you wanted to jump ahead, it should be something like:

```js
const reqOpts = extend(
  true,
  {
    uri: '',
    method: 'PATCH',
  },
  methodConfig.reqOpts,
  {
    json: metadata,
    qs: options,
  }
});
```

Reply:
Pending release: https://github.com/googleapis/nodejs-common/pull/452